File: data-structures/tree-traversals/java/bst_traverse_in_order.java
/*
Raspberry Pi Foundation
Developed as part of Ada Computer Science
Usage licensed under CC BY-NC-SA 4.0

Note: This file is designed to be copied out and compiled on your machine.
In order for the program to compile properly in some IDEs, you need to ensure that the
class filename is the same as the class name with the main method in it.

To run this file you need to:
1. Copy the contents
2. Create a Java project in the IDE of your choice (Eclipse, for example)
3. Create a new Java Class that uses the same name as the class containing the main method in this program
4. Paste the contents into the new Java Class
5. Save and run the program
*/

// A node in a binary search tree
class Node
{
    public int data;
    public Node left;
    public Node right;

    // Constructor method
    public Node(int data)
    {
        this.data = data;
    }
}


// A class for constructing a binary search tree (BST)
class BST
{
    private Node root; // Do not initialise the root yet as the tree is empty

    // Return the root node
    public Node getRoot()
    {
        return root;
    }


    // Insert a new node
    public void insert(int item)
    {
        // Create a new node
        Node newNode = new Node(item);

        // Check if the root node exists
        if (root == null) {
            // Set the root to be the new node
            root = newNode;
        }
        // Find the correct place to insert the new node
        else {
            Node current = root;
            boolean placed = false;

            // Repeat while the data has not been inserted
            while (placed == false) {
                // Check if the new item is greater than the current node data
                if (item > current.data) {
                    // Check if the current node does not have a right child node
                    if (current.right == null) {
                        // Insert the new node to the right of the current node
                        current.right = newNode;
                        placed = true;
                    }
                    // Otherwise repeat with the current right node
                    else {
                        current = current.right;
                    }
                }

                // Otherwise the new item is less than or equal to the current node
                else {
                    // Check if the current node does not have a left child node
                    if (current.left == null) {
                        // Insert the new node to the left of the current node
                        current.left = newNode;
                        placed = true;
                    }
                    // Otherwise repeat with the current left node
                    else {
                        current = current.left;
                    }
                }
            }
        }
    }


    // An in-order traversal of the binary search tree
    public void inOrderTraversal(Node node)
    {
        // Check any nodes to the left of the current node
        if (node.left != null) {
            inOrderTraversal(node.left);
        }

        // Output the data of the current node
        System.out.println(node.data);

        // Check any nodes to the right of the current node
        if (node.right != null) {
            inOrderTraversal(node.right);
        }
    }


}


class Testing
{
    // The main method is the entry point for all Java programs
    public static void main(String[] args)
    {
        // Instantiate a new tree object
        BST bst = new BST();

        // Insert test data into the binary search tree
        insertTestData(bst);

        // Get the root node
        Node root = bst.getRoot();

        // Testing
        System.out.println("### Output the tree structure ###");
        System.out.println("The root is the left-most value and the children are to the right\n");
        outputTree(root, 0);

        // Output the binary search tree using an in-order traversal
        System.out.println("\n### Binary search tree (in-order traversal) ###");
        bst.inOrderTraversal(root);
    }


    // Insert test data into the binary search tree
    public static void insertTestData(BST bst)
    {
        bst.insert(8);  // This will be the root node
        bst.insert(3);
        bst.insert(6);
        bst.insert(10);
        bst.insert(1);
        bst.insert(5);
        bst.insert(18);
        bst.insert(7);
        bst.insert(12);
    }


    // Output the tree with the root to the left and children to the right
    public static void outputTree(Node node, int level)
    {
        if (node != null) {
            outputTree(node.right, level + 1);

            String spaces = ""; // String of spaces
            if (level > 0) {
                int numSpaces = 4 * level;
                spaces = String.format("%-" + numSpaces + "s", "");
            }
            System.out.println(spaces + "-> " + node.data);

            outputTree(node.left, level + 1);
        }
    }


}

File: data-structures/tree-traversals/visual-basic/bst_traverse_in_order.vb
' Raspberry Pi Foundation
' Developed as part of Ada Computer Science
' Usage licensed under CC BY-NC-SA 4.0
' Note: This file is designed to be copied out and compiled on your machine.
' To run this file you need to:
' 1. Copy the contents
' 2. Paste them into the Visual Basic IDE of your choice (Visual Studio, for example)
' 3. Compile the program
' 4. Run the program


' A class to represent a node in a binary search tree
Class Node
    Private data As Integer ' Tree is storing integer values
    Private left As Node
    Private right As Node

    ' Constructor method
    Public Sub New(ByVal data As Integer)
        Me.data = data
    End Sub

    Public Function GetData() As Integer
        Return data
    End Function

    Public Function GetLeft() As Node
        Return left
    End Function

    Public Sub SetLeft(ByVal newLeft As Node)
        left = newLeft
    End Sub

    Public Function GetRight() As Node
        Return right
    End Function

    Public Sub SetRight(ByVal newRight As Node)
        right = newRight
    End Sub
End Class


' A class for constructing a binary search tree (BST)
Class BST
    Private root As Node ' Do not intialise the root yet as the tree is empty

    ' Return the root node
    Public Function GetRoot() As Node
        Return root
    End Function

    ' Insert a new node
    Public Sub Insert(ByVal item As Integer)
        ' Create a new node
        Dim newNode As Node = New Node(item)

        ' Check if the root node exists
        If root Is Nothing Then
            ' Set the root to be the new node
            ' Find the correct place to insert the new node
            root = newNode
        Else
            Dim current As Node = root
            Dim placed As Boolean = False

            ' Repeat while the data has not been inserted
            While placed = False
                ' Check if the new item is greater than the current node data
                If item > current.GetData() Then
                    ' Check if the current node does not have a right child node
                    If current.GetRight() Is Nothing Then
                        ' Insert the new node to the right of the current node
                        current.SetRight(newNode)
                        placed = True
                    ' Otherwise repeat with the current right node
                    Else
                        current = current.GetRight()
                    End If
                
                ' Otherwise the new item is less than or equal to the current node
                Else
                    ' Check if the current node does not have a left child node
                    If current.GetLeft() Is Nothing Then
                        ' Insert the new node to the left of the current node
                        current.SetLeft(newNode)
                        placed = True
                    ' Otherwise repeat with the current left node
                    Else
                        current = current.GetLeft()
                    End If
                End If
            End While
        End If
    End Sub


    ' An in-order traversal of the binary search tree
    Public Sub InOrderTraversal(ByVal node As Node)
        ' Check any nodes to the left of the current node
        If node.GetLeft() IsNot Nothing Then
            InOrderTraversal(node.GetLeft())
        End If

        ' Output the data of the current node
        Console.WriteLine(node.GetData())

        ' Check any nodes to the right of the current node
        If node.GetRight() IsNot Nothing Then
            InOrderTraversal(node.GetRight())
        End If
    End Sub
End Class


Module program

    ' The Main method is the entry point for all VB programs
    Sub Main()
        ' Instantiate a new tree object
        Dim bst As BST = New BST()

        ' Insert test data into the binary search tree
        InsertTestData(bst)

        ' Get the root node
        Dim root As Node = bst.GetRoot()

        ' Testing
        Console.WriteLine("### Output the tree structure ###")
        Console.WriteLine("The root is the left-most value and the children are to the right" & vbLf)
        OutputTree(root)

        ' Output the binary search tree using an in-order traversal
        Console.WriteLine(vbLf & "### Binary search tree (in-order traversal) ###")
        bst.InOrderTraversal(root)
    End Sub


    ' Insert test data into the binary search tree
    Sub InsertTestData(ByVal bst As BST)
        bst.Insert(8)  ' This will be the root node
        bst.Insert(3)
        bst.Insert(6)
        bst.Insert(10)
        bst.Insert(1)
        bst.Insert(5)
        bst.Insert(18)
        bst.Insert(7)
        bst.Insert(12)
    End Sub


    ' Output the tree with the root to the left and children to the right
    Sub OutputTree(ByVal node As Node, ByVal Optional level As Integer = 0)
        If node IsNot Nothing Then
            OutputTree(node.GetRight(), level + 1)

            Dim spaces As String = New String(" "c, 4 * level) ' String of spaces
            
            Console.WriteLine($"{spaces}-> {node.GetData()}")
            OutputTree(node.GetLeft(), level + 1)
        End If
    End Sub


End Module
